lib/_http_client.js

```diff
@@ -819,11 +793,30 @@ ClientRequest.prototype.onSocket = function onSocket(socket, err) {
 };
 
 function onSocketNT(req, socket, err) {
-  if (req.destroyed) {
-    _destroy(req, socket, req[kError]);
-  } else if (err) {
+  if (req.destroyed || err) {
     req.destroyed = true;
-    _destroy(req, null, err);
+
+    function _destroy(req, err) {
+      if (!req.aborted && !err) {
+        err = connResetException('socket hang up');
+      }
+      if (err) {
+        req.emit('error', err);
+      }
+      req._closed = true;
+      req.emit('close');
+    }
+
+    if (!err && req.agent) {
+      socket?.emit('free');
+    } else if (socket) {
+      finished(socket.destroy(err || req[kError]), (er) => {
+        _destroy(req, er || err);
+      });
+      return;
+    }
+
+    _destroy(req, err || req[kError]);
```

Comment:
A bunch of these can be slightly niftier code-style wise (`??`s etc) but that's pretty dumb to block on.